flowmachine/flowmachine/core/model_result.py

```diff
@@ -150,25 +152,50 @@ def to_sql(
 
         def do_query() -> ModelResult:
             logger.debug("Getting storage lock.")
-            with rlock(self.redis, self.md5):
+            q_state_machine = QueryStateMachine(self.redis, self.md5)
+            current_state, this_thread_is_owner = q_state_machine.execute()
+            if this_thread_is_owner:
                 logger.debug("Obtained storage lock.")
                 con = self.connection.engine
-                if force:
-                    self.invalidate_db_cache(name, schema=schema)
                 try:
                     with con.begin():
                         logger.debug("Using pandas to store.")
                         self._df.to_sql(name, con, schema=schema, index=False)
+                        # Mark as finished before writing meta to avoid getting blocked when
+                        # meta calls _make_query
+                        q_state_machine.finish()
                         if schema == "cache":
                             self._db_store_cache_metadata(compute_time=self._runtime)
+
                 except AttributeError:
                     logger.debug(
                         "No dataframe to store, presumably because this"
                         " was retrieved from the db."
                     )
-            logger.debug("Released storage lock.")
+            elif q_state_machine.is_executing:
+                logger.debug(
+                    f"Model result '{self.md5}' being written elsewhere, waiting for it to finish."
+                )
+                while q_state_machine.is_executing:
+                    sleep(5)
+
+            if q_state_machine.is_completed:
+                return self
+            elif q_state_machine.is_cancelled:
+                logger.error(f"Model result write '{self.md5}' was cancelled.")
+                raise QueryCancelledException(self.md5)
+            elif q_state_machine.is_errored:
+                logger.error(f"Model result write '{self.md5}' finished with an error.")
+                raise QueryErroredException(self.md5)
+                logger.debug("Released storage lock.")
```

Comment:
Where exactly is the storage lock released, and why does this only happen in the `is_errored` branch?